### PR TITLE
Avoid time domain casting exception during request cancellation

### PR DESCRIPTION
**Pull request by johnwheffner**
Context::Cancel() sets the deadline to DateTime::min(), which throws a range
exception when cast to system_clock::time_point.  Instead, cast
system_clock::now() to DateTime.

### CODE REVIEW

File: sdk/storage/azure-storage-common/src/storage_per_retry_policy.cpp
Comment (ahsonkhan) [inline sdk/storage/azure-storage-common/src/storage_per_retry_policy.cpp:44]: Why -1 as the alternative sentinel?

**Comment by ahsonkhan** [comment]
> Context::Cancel() sets the deadline to DateTime::min(), which throws a range
exception when cast to system_clock::time_point. Instead, cast
system_clock::now() to DateTime.

This issue might be related or at least provide some context:
https://github.com/Azure/azure-sdk-for-cpp/issues/2038

I am not sure if we want to do this, but if `Context::Cancel()` set the value to `DateTime::now()`, would that help address the issue?

File: sdk/storage/azure-storage-common/src/storage_per_retry_policy.cpp
Comment (johnwheffner) [reply sdk/storage/azure-storage-common/src/storage_per_retry_policy.cpp:44]: It's not a sentinel; the here is to protect against underflow.  In this case, any non-positive value is as good as another.

**Comment by johnwheffner** [comment]
> I am not sure if we want to do this, but if `Context::Cancel()` set the value to `DateTime::now()`, would that help address the issue?

It would solve this particular issue.  Seems to me there are other drawbacks to that approach, especially as you're not guaranteed a steady clock.  But this _does_ seem likely this kind of bug will crop up again without a design change.

**Review by Jinming-Hu — APPROVED** [review]

**Comment by vhvb1989** [comment]
@Jinming-Hu  do we need changelog for this change?  is it a breaking change?  I leave it up to you to click the Merge button for this PR.

**Pull request merged**
